gitpod-io/gitpod
# [server] Restrict snapshot access based on repository access

## Description


Restrict snapshot access based on repository access.

Also refactors:
- `GuardedSnapshot.workspace` (simplified)
- `WorkspaceLogAccessGuard` → `RepositoryResourceGuard`
- `RepositoryService.canAccessHeadlessLogs` → `RepositoryProvider.hasReadAccess`

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/8257

## How to test


1. Create a snapshot for a private repo
2. Verify that you can open the snapshot
3. Verify that other members of the private repo can open the snapshot
4. Verify that non-members of the private repo can **not** open the snapshot

Also:

1. Create a snapshot for a public repo
2. Verify that anyone can open the snapshot (even if they haven't connected the repo's Git provider)

## Release Notes

```release-note
Restrict snapshot access based on repository access
```

## Documentation
